### PR TITLE
implement `INTERSECT` and `EXCEPT` clause

### PR DESCRIPTION
**Pull request by JCOR11599**
This PR implements two MySQL functions, [`INTERSECT`](https://dev.mysql.com/doc/refman/8.0/en/intersect.html) and [`EXCEPT`](https://dev.mysql.com/doc/refman/8.0/en/except.html).
They are similar to `UNION` in that they combine to two sets of rows.
`INTERSECT` is the set of rows that are in both left and right.
`EXCEPT` is the set of rows that are in the left that are not in the right.

Additionally, there is a bug with field indexes with ORDER BY over unions.

companion pr: https://github.com/dolthub/vitess/pull/271

fixes https://github.com/dolthub/dolt/issues/6643

### CODE REVIEW

File: sql/rowexec/rel_iters.go [hunk 1]
```diff
@@ -1068,3 +1064,135 @@ func (ui *unionIter) Close(ctx *sql.Context) error {
 		return nil
 	}
 }
+
+type intersectIter struct {
+	lIter, rIter sql.RowIter
+	cached       bool
+	cache        map[uint64]int
```
Comment (max-hoffman) [inline sql/rowexec/rel_iters.go:1071]: tests with these nested as join RHS will stress the caching, which will also maybe expose a memory leak in the Dispose() cleanup of these

File: sql/plan/union.go [hunk 2]
```diff
@@ -21,13 +21,20 @@ import (
...
 )
 
+const (
+	UnionType = iota
+	IntersectType
+	ExceptType
+)
+
 // Union is a node that returns everything in Left and then everything in Right
 type Union struct {
```
Comment (max-hoffman) [inline sql/plan/union.go:31]: - we'll need something more generic to represent the collection of these. `SetArithmeticOp`? `SetArithmeticType[Union,Intersect,Set]`
-  i am also a bit worried about confusing some of the RECURSIVE CTE rules. for example, should `WITH RECURSIVE CTE UNION ... INTERSECT ... EXCEPT ... UNION` treat all of those nodes equivalently?

File: enginetest/queries/script_queries.go [hunk 3]
```diff
@@ -138,6 +138,125 @@ var ScriptTests = []ScriptTest{
...
+				Query: "table t1 intersect table t2;",
+				Expected: []sql.Row{
+					{1},
+					{3},
+				},
+			},
+			{
+				// Field indexing error
+				Skip:  true,
+				Query: "table t1 intersect table t2 order by i;",
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply enginetest/queries/script_queries.go:196]: I'm seeing that Union is now `SetArithemticOp`, so I'm less certain about the cause now. I thought we were trying to index `order by i` with `[i,i]`. We pick the closest match from the right. But `GetField(1)` is invalid, and will error. We should fix either way before shipping

File: enginetest/queries/script_queries.go [hunk 4]
```diff
@@ -138,6 +138,125 @@ var ScriptTests = []ScriptTest{
 			},
 		},
 	},
+	{
+		Name: "intersection tests",
```
Comment (max-hoffman) [inline enginetest/queries/script_queries.go:142]: i noted below that RECURSIVE CTE should be its own testing block, because of the way we are overloading the nodes, and I don't know whether MySQL supports including INTERSECT/EXCEPT in their special union handling.

Comment (max-hoffman) [reply enginetest/queries/script_queries.go:142]: need join testing. joins with INTERSECT/EXCEPT as both LHS and RHS. Joins with subqueries that have INTERSECT/EXCEPT. ideally in join_op_tests, so that we run them with all of the execution operators

Comment (max-hoffman) [reply enginetest/queries/script_queries.go:142]: nested EXCEPT/INTERCEPT seems like an initial layer of testing we have to have also. plus set arith in subqueries and subquery expressions to make sure we index them correctly

Comment (max-hoffman) [reply enginetest/queries/script_queries.go:142]: not seeing LIMIT, GROUP BY, HAVING tests

Comment (max-hoffman) [reply enginetest/queries/script_queries.go:142]: maybe some lateral join tests

**Review by max-hoffman — COMMENTED** [review]
Listed some suggestions, my biggest worry is probably indexing bugs with joins/subqueries, memory leaks with the new caches, and RECURSIVE CTE weirdness

File: enginetest/queries/script_queries.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (JCOR11599) [reply enginetest/queries/script_queries.go:196]: Similar to `INTERSECT` and `EXCEPT`, `UNION` along with `ORDER BY` also fail in the same way. I suspect it has something to do with being unable to find an exact column match since the types are different. 

Will continue investigating

File: enginetest/join_op_tests.go [hunk 5]
```diff
@@ -203,6 +203,39 @@ var joinOpTests = []struct {
 				Query:    "select * from xy where x = 1 and x in (select y from xy union select 1)",
 				Expected: []sql.Row{{1, 1}},
 			},
+			{
+				Query: "select * from xy where x = 1 intersect select * from uv;",
```
Comment (max-hoffman) [inline enginetest/join_op_tests.go:207]: could you do one with each with the format i have on line 203? in that gets converted into semi join?

File: sql/plan/recursive_cte.go [hunk 6]
```diff
@@ -43,7 +43,7 @@ import (
 // projection count and types. [Init] will be resolved before
 // [Rec] or [RecursiveCte] to share schema types.
 type RecursiveCte struct {
-	union *Union
+	setOp *SetOp
```
Comment (nicktobey) [inline sql/plan/recursive_cte.go:46]: RecursiveCTEs can only be unions, right? So the names in this file should probably reflect that. (And we should make sure that the SetOp has its op type set to Union.

File: sql/plan/set_op.go [hunk 7]
```diff
@@ -0,0 +1,249 @@
+// Copyright 2020-2021 Dolthub, Inc.
```
Comment (nicktobey) [inline sql/plan/set_op.go:1]: If you use `git mv`, you can make git see that this is a rename of `union.go`

File: sql/plan/concat.go [hunk 8]
```diff
@@ -20,7 +20,7 @@ import (
 
 // Concat is a node that returns everything in Left and then everything in
 // Right, but it excludes any results in Right that already appeared in Left.
-// Similar to Distinct(Union(...)) but allows Left to return the same row
+// Similar to Distinct(SetOp(...)) but allows Left to return the same row
```
Comment (nicktobey) [inline sql/plan/concat.go:23]: This should probably still say Union

File: sql/plan/recursive_cte.go [hunk 9]
```diff
@@ -90,7 +91,7 @@ func (r *RecursiveCte) Right() sql.Node {
 	return r.union.right
 }
 
-func (r *RecursiveCte) Union() *Union {
+func (r *RecursiveCte) SetOp() *SetOp {
```
Comment (nicktobey) [inline sql/plan/recursive_cte.go:94]: This should still be called Union.

File: sql/planbuilder/cte.go [hunk 10]
```diff
@@ -188,28 +206,28 @@ func (b *Builder) buildRecursiveCte(inScope *scope, union *ast.Union, name strin
 // "should have one or more non-recursive query blocks followed by one or more recursive ones"
 // "the recursive table must be referenced only once, and not in any subquery"
 func splitRecursiveCteUnion(name string, n ast.SelectStatement) (ast.SelectStatement, ast.SelectStatement) {
-	union, ok := n.(*ast.Union)
```
Comment (nicktobey) [inline sql/planbuilder/cte.go:191]: I think the variable in this function should still be called `union`. After all, we know it's a union.

File: sql/rowexec/rel.go [hunk 11]
```diff
@@ -644,50 +644,66 @@ func (b *BaseBuilder) buildIndexedTableAccess(ctx *sql.Context, n *plan.IndexedT
 	return sql.NewSpanIter(span, sql.NewTableRowIter(ctx, n.Table, partIter)), nil
 }
 
-func (b *BaseBuilder) buildUnion(ctx *sql.Context, u *plan.Union, row sql.Row) (sql.RowIter, error) {
-	span, ctx := ctx.Span("plan.Union")
-	var iter sql.RowIter
+func (b *BaseBuilder) buildUnion(ctx *sql.Context, s *plan.SetOp, row sql.Row) (sql.RowIter, error) {
```
Comment (nicktobey) [inline sql/rowexec/rel.go:647]: This should probably be called buildSetOp now.

File: sql/rowexec/rel.go [hunk 12]
```diff
@@ -644,50 +644,66 @@ func (b *BaseBuilder) buildIndexedTableAccess(ctx *sql.Context, n *plan.IndexedT
...
+	}
+	iter2, err = b.buildNodeExec(ctx, s.Right(), row)
 	if err != nil {
 		span.End()
 		return nil, err
 	}
-	iter = &unionIter{
-		cur: iter,
-		nextIter: func(ctx *sql.Context) (sql.RowIter, error) {
-			return b.buildNodeExec(ctx, u.Right(), row)
```
Comment (nicktobey) [inline sql/rowexec/rel.go:660]: So I think the reason why `nextIter` was a provider function was so that if the engine doesn't exhaust the first rowIter in a union, it doesn't need to initialize the second. We probably want to preserve that behavior.

File: sql/plan/set_op.go (hunk 7 above)
Comment (JCOR11599) [reply sql/plan/set_op.go:1]: Just tried, it didn't do it

File: sql/rowexec/rel.go [hunk 13]
```diff
@@ -644,50 +644,66 @@ func (b *BaseBuilder) buildIndexedTableAccess(ctx *sql.Context, n *plan.IndexedT
...
+		}
+	case plan.ExceptType:
+		iter = &exceptIter{
+			lIter: iter,
+			rIter: iter2,
+		}
 	}
-	if u.Distinct {
+
+	if s.Distinct {
```
Comment (nicktobey) [inline sql/rowexec/rel.go:679]: Be careful: This works for `UNION` and `INTERSECT`, but not `EXCEPT`. `DISTINCT(TABLE a EXCEPT ALL TABLE b)` is not the same as `TABLE a EXCEPT DISTINCT TABLE b`.

`EXCEPT ALL` subtracts occurrences in the right table from the left table. So a value that appears three times on the left and once on the right would appear twice in the output.

`EXCEPT DISTINCT` computes the set difference: if a row appears on the right, it will never appear in the output, regardless of how many times it appears on the left. 

Thus, the following have different results:

```
mysql> SELECT DISTINCT * FROM (VALUES ROW(1), ROW(1), ROW(1) EXCEPT ALL VALUES ROW(1)) sub_expr;
+----------+
| column_0 |
+----------+
|        1 |
+----------+
1 row in set (0.00 sec)

mysql> VALUES ROW(1), ROW(1), ROW(1) EXCEPT DISTINCT VALUES ROW(1);
Empty set (0.00 sec)
```

It should be correct to insert the DistinctIters into the *inputs* to the Except:

```
mysql> SELECT DISTINCT * FROM (VALUES ROW(1), ROW(1), ROW(1)) sub_expr_l EXCEPT ALL SELECT DISTINCT * FROM (VALUES ROW(1)) sub_expr_r;
Empty set (0.00 sec)
```

File: sql/plan/set_op.go (hunk 7 above)
Comment (max-hoffman) [reply sql/plan/set_op.go:1]: might work if you do it inside a to-main squash

File: sql/rowexec/rel.go (hunk 12 above)
Comment (max-hoffman) [reply sql/rowexec/rel.go:660]: make sense to me

File: sql/rowexec/rel.go (hunk 13 above)
Comment (JCOR11599) [reply sql/rowexec/rel.go]: Good catch, I fixed the logical error.

However, we don't actually parse `(VALUES ROW(1), ROW(1), ...)` here and in many other places.
I wrote a test with the same values in tables. I think a fix for parsing `values_statement` would go in a separate PR.

File: enginetest/memory_engine_test.go [hunk 14]
```diff
@@ -190,18 +190,12 @@ func TestSingleScript(t *testing.T) {
 		{
 			Name: "insert default value",
 			SetUpScript: []string{
-				"create table t1 (i int);",
```
Comment (nicktobey) [inline enginetest/memory_engine_test.go]: Revert the changes to this file.

**Review by nicktobey — APPROVED** [review]
Looks good to me!
